### PR TITLE
rgw: civetweb update to 1.10

### PR DESCRIPTION
**Pull request by theanalyst**
wip branch for civetweb 1.10 migration

Reverted the following downstream patches and reapplied a few:
* b6ab7abac864d243e24785e5a110f418dc0174d2 - Build fix, NO_SSL_DL; seems to be present in the current civetweb.c
* f8504eff9e745c86a577f4207341577c36312f05 - introduces mg_get_local_addr: reapplied as 9aae1128b7bf9537d353891edc58b14d8f4e66a1
* 09c580e9aa11a4e75ceaebde8ae18b6e0d43bf9e: return 400 instead of 411 for invalid headers? reapplied as 302db2ddb5644ff03e51fb699a3c639a7768a1d0 cc @yehudasa 
* 3097b6975c40e7255edbe6b488a6b46642483f5c: introduces following enhancements, 
- canonicalize url path, 
- get_request_len 
- mg_set_http_status
reapplied in c598009e253cba9b8d2944faac33f28d343f9cf7

The following while reverted for easier merge; are already part of current upstream and this changeset 
* 5da61b9dffbe7e7d4667880f2ca136cb39680839 & a24ddf60bb1b0467121e26261d63da0548a4b3c1: downstream pr: https://github.com/ceph/civetweb/pull/16 which was already accepted upstream as https://github.com/civetweb/civetweb/pull/397
* b39a1f20d6f42fa0a8ef3e77e810c81044da5d02 transfer encoding takes precedence over content-length, downstream pr https://github.com/ceph/civetweb/pull/22, merged upstream as https://github.com/civetweb/civetweb/pull/511 
* e8b60cd716b05dd3556f171c976e12efc39c605d: discard unread data on read; already upstream as https://github.com/civetweb/civetweb/pull/362

### CODE REVIEW

File: src/civetweb.c [hunk 1]
```diff
@@ -5906,8 +8135,10 @@ connect_socket(struct mg_context *ctx /* may be NULL */,
 		return 0;
 	}
 
-#ifndef NO_SSL_DL
-	if (use_ssl && (SSLv23_client_method == NULL)) {
```
Comment (theanalyst) [inline src/civetweb.c]: earlier downstream commit b6ab7abac864d243e24785e5a110f418dc0174d2 which adds the ifndef NO_SSL_DL seems to be considered in the upstream changes

**Comment by cbodley** [comment]
this is probably a good time to fix the branch naming.. once this is ready, can we add a ceph-master branch that tracks the ceph master branch?

**Comment by cbodley** [comment]
@theanalyst so is this effectively the civetweb upstream branch with those two `downstream: ` commits on top?  can you just push that as a new ceph-master branch instead of merging it into wip-listen4?

**Comment by theanalyst** [comment]
@mdw-at-linuxbox @cbodley I additionally added a downstream feature called allow_unicode_in_urls a95285e02117998638cb1e2541fec3fd7c2c23e0 as civetweb was enforcing strict url checking, now the tests seem to be green at https://github.com/ceph/ceph/pull/21123, just did a cosmetic rebase reordering the commits

**Comment by theanalyst** [comment]
Changelog:
* fixed branch naming in civetweb:
ceph-jewel ceph-luminous and ceph-master introduced, which track the respective release branch sha1s now. Re targetted this pr against ceph-master

* reworded all the downstream commits with a downstream prefix
we have 4 downstream commits now, it is easier to revert/reapply as necessary during future upgrades

**Review by cbodley — APPROVED** [review]

**Pull request merged**